### PR TITLE
Skytap template relationship refactoring

### PR DESCRIPTION
**Pull request by mhgoldman**
Add state name to skytap_template and create has_many relationship from course_region

### CODE REVIEW

File: app/models/course_region.rb [hunk 1]
```diff
@@ -7,11 +7,25 @@ class CourseRegion < ActiveRecord::Base
...
+  end
+
+  def skytap_template_for_state(state_name)
+    skytap_templates.where("lower(state_name) = ?", state_name.to_s.downcase).first
+  end
+
+  #TODO legacy cruft, refactor me out
+  def default_state_skytap_template=(template)
+    template.course_region = self
+    template.state_name = ''
```
Comment (eliotsykes) [inline app/models/course_region.rb]: On lines 23 and 13, consider introducing an explaining constant for the empty string, e.g. Skytap::Template::DEFAULT_STATE_NAME.

File: app/models/course_region.rb [hunk 2]
```diff
@@ -7,11 +7,25 @@ class CourseRegion < ActiveRecord::Base
 
   has_many :participants
 
-  has_one :skytap_template, class_name: 'Skytap::Template'
-    
+  has_many :skytap_templates, class_name: 'Skytap::Template'
+  
+  def default_state_skytap_template
+    t = skytap_templates.find_by(state_name: '') || skytap_templates.find_by(state_name: nil)
```
Comment (eliotsykes) [inline app/models/course_region.rb]: Consider migrating the `state_name` column to be non-nullable so the 2nd `find_by ...` can be omitted from line 13.

File: app/models/skytap/template.rb [hunk 3]
```diff
@@ -26,12 +28,18 @@ def self.pull
       api_templates.each do |api_template|
         next if api_template.get.publish_sets.blank?
 
-        template = Skytap::Template.with_course_type_scope.where(courses: { course_type: course_type }).find_or_create_by(skytap_id: api_template.id)
+        course_name, state_name = api_template.name.split('|').map(&:strip)
+
+        template = Skytap::Template.with_course_type_scope.where(courses: { course_type: course_type }).find_or_initialize_by(skytap_id: api_template.id)
+        template.state_name = state_name
+        template.save if template.changed?
```
Comment (eliotsykes) [inline app/models/skytap/template.rb]: Line 35: If `template.save` fails (e.g. due to validation error), method execution will continue. Consider if one of these is more appropriate:

- `template.save!` raises error on failure
- Checking the result of `template.save`
- Checking `template.persisted?` afterward

File: app/models/skytap/template.rb [hunk 4]
```diff
@@ -26,12 +28,18 @@ def self.pull
...
+
         region = template.course_region.try(:region) || Region.find_or_create_by(name: api_template.region)
-        course = template.course_region.try(:course) || Course.find_or_create_by(name: api_template.name, course_type: course_type)
+        course = template.course_region.try(:course) || Course.find_or_create_by(name: course_name, course_type: course_type)
```
Comment (eliotsykes) [inline app/models/skytap/template.rb]: Consider `find_or_create_by!` on lines 37 & 38. `find_or_create_by` (no `!`) will fail silently but still return the unpersisted object if there are validation errors.

File: app/models/skytap/template.rb [hunk 5]
```diff
@@ -26,12 +28,18 @@ def self.pull
...
 
-        course_region = CourseRegion.find_or_create_by(region: region, course: course)
-        course_region.update(skytap_template: template) unless course_region.skytap_template == template
+        course_region = CourseRegion.find_or_initialize_by(region: region, course: course)
+        course_region.skytap_templates << template
+        course_region.save if course_region.changed?
```
Comment (eliotsykes) [inline app/models/skytap/template.rb]: As with above, consider what ought to happen if `.save` fails.

File: db/migrate/20160411191936_add_state_name_to_skytap_templates.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddStateNameToSkytapTemplates < ActiveRecord::Migration
+  def change
+    add_column :skytap_templates, :state_name, :string
+  end
+end
```
Comment (eliotsykes) [inline db/migrate/20160411191936_add_state_name_to_skytap_templates.rb]: Apologies if you're already planning to do this or discovered a problem with it. Consider writing a migration to add a multi-column uniqueness constraint on course_region_id+state_name columns.

File: db/seeds.rb [hunk 7]
```diff
@@ -23,7 +23,7 @@
   regions[0..1].each do |region|
     courses.each do |course|
       course_region = CourseRegion.find_or_initialize_by(region: region, course: course)
-      course_region.create_skytap_template if course_region.skytap_template.nil?
+      course_region.create_skytap_template if course_region.default_state_skytap_template.nil?
```
Comment (eliotsykes) [inline db/seeds.rb]: `CourseRegion#create_skytap_template` is provided by the original `has_one :skytap_template` association.

With `has_many :skytap_templates`, `CourseRegion#create_skytap_template` is no longer provided. Its replaced with `CourseRegion#skytap_templates#create`. Consider using `course_region.skytap_templates.create` here.

File: spec/models/skytap/configuration_spec.rb [hunk 8]
```diff
@@ -105,15 +104,18 @@ def create_provisioned_configuration
...
-    region = Region.find_or_create_by(name: 'US-East')
+    region = Region.find_or_create_by!(name: 'US-East')
     course = create(:course, name: 'PubURL Course')
     skytap_template = create(:skytap_template, skytap_id: 725173)
-    course_region = create(:course_region, region: region, course: course, skytap_template: skytap_template)
+    course_region = create(:course_region, region: region, course: course)
+    course_region.skytap_templates.destroy_all
+    course_region.skytap_templates << skytap_template
+    skytap_template.save!
     event = create(:event, course: course, name: 'Nifty course event', skytap_project: build(:skytap_project, skytap_id: 56996))
```
Comment (eliotsykes) [inline spec/models/skytap/configuration_spec.rb]: Consider enhancing course_region factory so lines 110-113 can be reduced to something like:

```diff
- course_region = create(:course_region, region: region, course: course)
- course_region.skytap_templates.destroy_all
- course_region.skytap_templates << skytap_template
- skytap_template.save!
+ course_region = create(:course_region, region: region, course: course, skytap_templates: [skytap_template])
```

File: spec/models/skytap/template_spec.rb [hunk 9]
```diff
@@ -179,8 +184,8 @@ def expect_initial_templates
       course = Course.find_by!(name: course_item[:name], course_type: course_item[:type])
       course_item[:templates].each do |region_name, skytap_id|
         region = Region.find_by!(name: region_name)
+
         course_region = course.course_regions.find_by!(region: region)
-        expect(course_region.skytap_template.skytap_id).to eq(skytap_id)
```
Comment (eliotsykes) [inline spec/models/skytap/template_spec.rb]: `skytap_id`s don't seem to be used in `expect_initial_templates`. If this is intended, consider removing the skytap ids from `course_items` test data above so its clear they're not needed and there's no missing test coverage here.

**Comment by mhgoldman** [comment]
@eliotsykes Only the last 2 commits are new.

**Comment by eliotsykes** [comment]
:+1: thanks, looks good

**Pull request merged**